File: java/src/VideoLibrary.java
package src;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Scanner;
import java.util.stream.Collectors;

/**
 * A class used to represent a Video Library.
 */
class VideoLibrary {

  private final HashMap<String, Video> videos;

  VideoLibrary() {
    this.videos = new HashMap<String, Video>();
    try {
      Scanner scanner = new Scanner(new File("videos.txt"));
      while (scanner.hasNextLine()) {
        String line = scanner.nextLine();
        String[] split = line.split("\\|");
        String title = split[0].strip();
        String id = split[1].strip();
        List<String> tags = Arrays.stream(split[2].split(",")).map(String::strip).collect(
            Collectors.toList());
        this.videos.put(id, new Video(title, id, tags));
      }

    } catch (FileNotFoundException e) {
      System.out.println("Couldn't find videos.txt");
      e.printStackTrace();
    }
  }

  List<Video> getVideos() {
    return new ArrayList<Video>(this.videos.values());
  }

  Video getVideo(String videoId) throws Exception {
    if (this.videos.containsKey(videoId)) {
      return this.videos.get(videoId);
    }
    throw new Exception("Video not found in video library");
  }
}

File: cpp/src/videolibrary.cpp
#include "videolibrary.h"

#include <fstream>
#include <iostream>
#include <sstream>
#include <unordered_map>
#include <vector>

#include "helper.h"
#include "video.h"

VideoLibrary::VideoLibrary() {
  std::ifstream file("./src/videos.txt");
  if (file.is_open()) {
    std::string line;
    while (std::getline(file, line)) {
      std::stringstream linestream(line);
      std::string title;
      std::string id;
      std::string tag;
      std::vector<std::string> tags;
      std::getline(linestream, title, '|');
      std::getline(linestream, id, '|');
      while (std::getline(linestream, tag, ',')) {
        tags.emplace_back(trim(std::move(tag)));
      }
      Video video = Video(trim(std::move(title)), trim(id), std::move(tags));
      this->videos.emplace(trim(std::move(id)), std::move(video));
    }
  } else {
    std::cout << "Couldn't find videos.txt" << std::endl;
  }
}

std::vector<Video> VideoLibrary::getVideos() const {
  std::vector<Video> result;
  for (auto const &video : this->videos) {
    result.emplace_back(video.second);
  }
  return result;
}

Video const *VideoLibrary::getVideo(std::string videoId) const {
  auto const found = this->videos.find(videoId);
  if (found == this->videos.end()) {
    std::cout << "Video not found in video library" << std::endl;
    return nullptr;
  } else {
    return &(found->second);
  }
}
